chore: trigger ecosystem-ci by comment

sapphi-red opened the pull request:
### Description
Make it possible to run ecosystem-ci by commenting `/ecosystem-ci run` to PR.

~~Example: https://github.com/sapphi-red/vite/pull/2
(the CI ran failed because I commented out Line 37 which requires https://github.com/vitejs/vite-ecosystem-ci/pull/120)~~

requires https://github.com/vitejs/vite-ecosystem-ci/pull/122

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [x] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

sapphi-red committed:
chore: trigger ecosystem-ci by comment

sapphi-red commented on .github/workflows/ecosystem-ci-trigger.yml:
```diff
@@ -0,0 +1,69 @@
+      - uses: actions/github-script@v6
+        id: trigger
+        with:
+          # this token needs "repo" scope
+          github-token: ${{ secrets.ECOSYSTEM_CI_TRIGGER_GITHUB_PAT }}
```
This needs to be set.

sapphi-red commented on .github/workflows/ecosystem-ci-trigger.yml:
```diff
@@ -0,0 +1,69 @@
+              }
+            })
+
+            await new Promise(resolve => setTimeout(resolve, 10 * 1000))
+
+            console.log(`Get lastest workflow run`)
+            // assumes the first one is the one triggered
+            const { data: runs } = await github.rest.actions.listWorkflowRuns({
+              owner: context.repo.owner,
+              repo: 'vite-ecosystem-ci',
+              workflow_id: 'ecosystem-ci.yml',
+              event: 'workflow_dispatch',
+              created: `>${startDate.toISOString()}`,
+              per_page: 1
+            })
```
There isn't a way to get which workflow run is the triggered one. The URL might be wrong when many workflow runs were triggered at the same time.

patak-dev replied:
Possible workaround: https://stackoverflow.com/a/69500478

sapphi-red commented on .github/workflows/ecosystem-ci-trigger.yml:
```diff
@@ -0,0 +1,69 @@
+jobs:
+  trigger-all:
+    runs-on: ubuntu-latest
+    if: github.event.issue.pull_request && startsWith(github.event.comment.body, '/ecosystem-ci run')
```
Should we limit who can trigger this? (e.g. team members, PR author)

dominikg replied:
Yes, ecosystem ci is very intense. Would limit to team, not even PR author

dominikg replied:
Maybe cli like arg too to be able to call single framework tests? eg `/ecosystem-ci svelte`

patak-dev replied:
I don't know if the complexity of limiting it to the team is going to be needed, given that only close contributors will know how this works. I think we could try to leave it open and see how people use it, if there is abuse, we could restrict it. Maybe we could avoid sending to Discord the runs that are triggered by `/ecosystem-ci` and instead answer back with a comment on the PR?

+1 for passing options to `/ecosystem-ci` (maybe in future PRs)

patak-dev replied:
We discussed this with @dominikg in Discord, and he strongly suggest going with a whitelist. I'm fine with it, I would prefer if there is a way to have a list that doesn't depend on giving access to the Vite repo though (ideally any framework author doing a PR to vite should be able to trigger it). But we can keep it between the team to start with, as there are many of us active on PRs daily.

sapphi-red committed:
chore: change branch

sapphi-red commented:
I've implemented these features:

- allow user list
- show result after run & disable discord webhook
- `/ecosystem-ci {suite}`

and also changed to use GitHub App instead of PAT

Example: https://github.com/sapphi-red/vite/pull/3

sapphi-red commented:
The current flow:

```mermaid
sequenceDiagram
  actor U as user
  participant vPR as vitejs/vite<br>PR
  actor vGA as vitejs/vite<br>GitHub Actions
  actor eGA as vitejs/vite-ecosystem-ci<br>GitHub Actions

  U ->> vPR: comment on PR
  vPR ->> vGA: trigger
  vGA ->> vPR: add :+1: reaction to comment
  Note over vPR,vGA: when triggered user is not allowed, add :-1: reaction to comment and end
  vGA ->> eGA: trigger "ecosystem-ci-from-pr"
  eGA ->> vPR: add comment
  eGA ->> eGA: run ecosystem-ci
  eGA ->> vPR: add result to comment

  vPR ->> U: gets result
```

To setup GitHub App, the following steps are required:

1. Create GitHub App
1. Grant permissions
  - Actions: Read and Write
  - Issues: Read and Write
  - Pull requests: Read and Write
1. Install GitHub App
  - vitejs/vite
  - vitejs/vite-ecosystem-ci
1. Set secrets for vitejs/vite and vitejs/vite-ecosystem-ci

patak-dev commented:
Amazing work @sapphi-red 😍, and thanks again for the detailed explanations.
I think we should move forward to test the flows, and we can keep iterating later! This is too good!

sapphi-red commented:
@patak-dev I don't have permission to [create GitHub Apps under the vitejs org](https://github.com/organizations/vitejs/settings/apps) and set secrets to vitejs/vite and vitejs/vite-ecosystem-ci.
Would you setup these?

patak-dev commented:
@sapphi-red all done 🙌🏼

sapphi-red commented:
@patak-dev Thanks!

sapphi-red commented:
/ecosystem-ci run

patak-dev commented:
I love the 👍🏼 reaction on your comment ❤️

sapphi-red commented:
It failed.
https://github.com/vitejs/vite/runs/8112624941?check_suite_focus=true#step:4:7

Maybe these values are different?🤔
It seems the values are empty or not a expected value. (Note that I changed secret names for each repository to make it easier to know what it is)
https://github.com/vitejs/vite/blob/7c4ab51e744747fd626b7316425f003bcb537c49/.github/workflows/ecosystem-ci-trigger.yml#L68-L69

sapphi-red commented:
/ecosystem-ci run
